Fabric.Authorization.Persistence.SqlServer/Stores/SqlServerPermissionStore.cs

```diff
@@ -0,0 +1,210 @@
+            }
+
+            permission.IsDeleted = true;
+            await _authorizationDbContext.SaveChangesAsync();
+        }
+
+        public async Task Update(Permission model)
+        {
+            var permission = await _authorizationDbContext.Permissions
+                .SingleOrDefaultAsync(p =>
+                    p.PermissionId == model.Id
+                    && !p.IsDeleted);
+
+            if (permission == null)
+            {
+                throw new NotFoundException<User>($"Could not find {typeof(Permission).Name} entity with ID {model.Id}");
+            }
+
+            model.ToEntity(permission);
+            _authorizationDbContext.Permissions.Update(permission);
+            await _authorizationDbContext.SaveChangesAsync();
+        }
+
+        public async Task<bool> Exists(Guid id)
+        {
+            var permission = await _authorizationDbContext.Permissions
+                .SingleOrDefaultAsync(p =>
+                    p.PermissionId == id
+                    && !p.IsDeleted);
+
+            return permission != null;
+        }
+
+        public Task<IEnumerable<Permission>> GetPermissions(string grain, string securableItem = null,
+            string permissionName = null)
+        {
+            var permissions = _authorizationDbContext.Permissions
+                .Include(p => p.SecurableItem)
+                .Where(p => string.Equals(p.Grain, grain, StringComparison.OrdinalIgnoreCase));
+
+            if (!string.IsNullOrWhiteSpace(securableItem))
+            {
+                permissions = permissions.Where(p => string.Equals(p.SecurableItem.Name, securableItem));
+            }
+
+            if (!string.IsNullOrWhiteSpace(permissionName))
+            {
+                permissions = permissions.Where(p => string.Equals(p.Name, securableItem));
+            }
+
+            return Task.FromResult(permissions.Select(p => p.ToModel()).AsEnumerable());
+        }
+
+        public async Task AddOrUpdateGranularPermission(GranularPermission granularPermission)
+        {
+            var idParts = SplitUserId(granularPermission.Id);
+
+            var user = await _authorizationDbContext.Users
+                .Include(u => u.UserPermissions)
+                .ThenInclude(up => up.Permission)
+                .SingleOrDefaultAsync(u => u.IdentityProvider.Equals(idParts[0], StringComparison.OrdinalIgnoreCase)
+                                           && u.SubjectId.Equals(idParts.Length > 1 ? idParts[1] : idParts[0], StringComparison.OrdinalIgnoreCase)
+                                           && !u.IsDeleted);
+
+            if (user == null)
+            {
+                throw new NotFoundException<User>($"Could not find {typeof(User).Name} User ID {granularPermission.Id}");
+            }
+
+            // remove all current permissions first and then replace them with the new set of permissions
+            var currentUserPermissions = user.UserPermissions.Where(up => !up.IsDeleted);
```

Comment:
Is there a better way to do this than removing and readding all the permissions? I am concerned that EF will generate a bunch of deletes/inserts where only a few or one is required.

Reply:
We don't have to solve this with this PR, but it might be something we want to look at in the future.

Reply:
Ok, I was starting to think about it...I was mimicking the CouchDB behavior for now, but I agree there is probably a better way to do and don't love the current approach.